File: Java/li2/plp/imperative2/Exemplo4.java
package li2.plp.imperative2;

import li2.plp.expressions1.util.TipoPrimitivo;
import li2.plp.expressions2.expression.ExpEquals;
import li2.plp.expressions2.expression.ExpNot;
import li2.plp.expressions2.expression.ExpSub;
import li2.plp.expressions2.expression.Id;
import li2.plp.expressions2.expression.ValorInteiro;
import li2.plp.expressions2.expression.ValorString;
import li2.plp.expressions2.memory.IdentificadorJaDeclaradoException;
import li2.plp.expressions2.memory.IdentificadorNaoDeclaradoException;
import li2.plp.expressions2.memory.VariavelJaDeclaradaException;
import li2.plp.expressions2.memory.VariavelNaoDeclaradaException;
import li2.plp.imperative1.command.Atribuicao;
import li2.plp.imperative1.command.ComandoDeclaracao;
import li2.plp.imperative1.command.IfThenElse;
import li2.plp.imperative1.command.SequenciaComando;
import li2.plp.imperative1.command.Skip;
import li2.plp.imperative1.command.Write;
import li2.plp.imperative1.declaration.DeclaracaoComposta;
import li2.plp.imperative1.declaration.DeclaracaoVariavel;
import li2.plp.imperative1.memory.ContextoCompilacaoImperativa;
import li2.plp.imperative1.memory.ListaValor;
import li2.plp.imperative2.command.ChamadaProcedimento;
import li2.plp.imperative2.command.ListaExpressao;
import li2.plp.imperative2.declaration.DeclaracaoParametro;
import li2.plp.imperative2.declaration.DeclaracaoProcedimento;
import li2.plp.imperative2.declaration.DefProcedimento;
import li2.plp.imperative2.declaration.ListaDeclaracaoParametro;
import li2.plp.imperative2.memory.ContextoExecucaoImperativa2;
import li2.plp.imperative2.memory.ProcedimentoNaoDeclaradoException;

/**
 * Exemplo4 - Executa o seguinte programa na linguagem li2:
 *
 * { var b = 3,
 *   proc escreveRecursivo (int a) {
 *     if (not (a == 0)) then {
 *       var x = 0; x := a - 1;
 *       write("Ola");
 *       call escreveRecursivo(x)
 *     } else skip
 *   };
 *   call escreveRecursivo(a)   <- ERRO: 'a' nao esta declarado no escopo externo!
 * }
 *
 * Resultado esperado: excecao de variavel nao declarada (IdentificadorNaoDeclarado ou similar)
 */
public class Exemplo4 {

    /**
     * Executa um programa li2, verificando tipos antes de rodar.
     * Imprime a saida produzida pelo write ou a excecao capturada.
     *
     * @param descricao texto descritivo do programa
     * @param prog      o programa a executar
     * @param ambiente  o ambiente de execucao (contexto imperativo2)
     */
    static void executar(String descricao, Programa prog,
                         ContextoExecucaoImperativa2 ambiente) {
        System.out.println("--------------------------------------------------");
        System.out.println("Programa: \n" + descricao + "\n");
        try {
            boolean bemTipado = prog.checaTipo(new ContextoCompilacaoImperativa(new ListaValor()));
            System.out.println("Bem tipado: " + bemTipado);
            if (bemTipado) {
                ListaValor saida = prog.executar(ambiente);
                System.out.print("Saida: ");
                while (saida.length() > 0) {
                    System.out.print(saida.getHead() + " ");
                    saida = (ListaValor) saida.getTail();
                }
                System.out.println();
            } else {
                System.out.println("Erro de tipo: execucao abortada.");
            }
        } catch (VariavelNaoDeclaradaException e) {
            System.out.println("[ERRO] Variavel nao declarada: " + e.getMessage());
        } catch (VariavelJaDeclaradaException e) {
            System.out.println("[ERRO] Variavel ja declarada: " + e.getMessage());
        } catch (ProcedimentoNaoDeclaradoException e) {
            System.out.println("[ERRO] Procedimento nao declarado: " + e.getMessage());
        } catch (IdentificadorNaoDeclaradoException e) {
            System.out.println("[ERRO] Identificador nao declarado: " + e.getMessage());
        } catch (IdentificadorJaDeclaradoException e) {
            System.out.println("[ERRO] Identificador ja declarado: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("[ERRO] Erro inesperado: " + e.getMessage());
        }
        System.out.println("--------------------------------------------------");
    }

    public static void main(String[] args) {

        // Identificadores
        Id idB = new Id("b");
        Id idA = new Id("a");  // parametro do procedimento — NAO declarado no escopo externo
        Id idX = new Id("x");
        Id idEscreveRecursivo = new Id("escreveRecursivo");

        // Declaracao de variavel do bloco externo: var b = 3
        DeclaracaoVariavel declB = new DeclaracaoVariavel(idB, new ValorInteiro(3));

        // Parametro formal do procedimento: int a
        DeclaracaoParametro paramA = new DeclaracaoParametro(idA, TipoPrimitivo.INTEIRO);
        ListaDeclaracaoParametro listaParamA = new ListaDeclaracaoParametro(paramA);

        // Condicao do if: not (a == 0)
        ExpNot condicao = new ExpNot(new ExpEquals(idA, new ValorInteiro(0)));

        // Corpo do bloco then:
        //   var x = 0; x := a - 1; write("Ola"); call escreveRecursivo(x)

        // Declaracao local: var x = 0
        DeclaracaoVariavel declX = new DeclaracaoVariavel(idX, new ValorInteiro(0));

        // Atribuicao: x := a - 1
        Atribuicao atribX = new Atribuicao(idX, new ExpSub(idA, new ValorInteiro(1)));

        // write("Ola")
        Write writeOla = new Write(new ValorString("Ola"));

        // call escreveRecursivo(x) — parametro real: x
        ChamadaProcedimento chamadaRecursiva = new ChamadaProcedimento(
                idEscreveRecursivo, new ListaExpressao(idX));

        // Sequencia dentro do then: x := a-1; write("Ola"); call escreveRecursivo(x)
        SequenciaComando seqThen = new SequenciaComando(
                atribX,
                new SequenciaComando(writeOla, chamadaRecursiva)
        );

        // Bloco then completo: { var x = 0; x := a-1; write("Ola"); call escreveRecursivo(x) }
        ComandoDeclaracao blocoThen = new ComandoDeclaracao(declX, seqThen);

        // if (not (a == 0)) then { ... } else skip
        IfThenElse ifCmd = new IfThenElse(condicao, blocoThen, new Skip());

        // Definicao do procedimento escreveRecursivo (int a) { if ... }
        DefProcedimento defEscreveRecursivo = new DefProcedimento(listaParamA, ifCmd);

        // Declaracao do procedimento: proc escreveRecursivo (int a) { ... }
        DeclaracaoProcedimento declEscreveRecursivo =
                new DeclaracaoProcedimento(idEscreveRecursivo, defEscreveRecursivo);

        // Declaracao composta do bloco externo: var b = 3, proc escreveRecursivo (int a) { ... }
        DeclaracaoComposta declaracoes = new DeclaracaoComposta(declB, declEscreveRecursivo);

        // Comando principal: call escreveRecursivo(a)
        // ATENCAO: 'a' nao esta declarado no escopo externo — deve gerar erro!
        ChamadaProcedimento chamadaInicial = new ChamadaProcedimento(
                idEscreveRecursivo, new ListaExpressao(idA));

        // Bloco completo:
        // { var b = 3, proc escreveRecursivo (int a) { ... }; call escreveRecursivo(a) }
        ComandoDeclaracao blocoCompleto = new ComandoDeclaracao(declaracoes, chamadaInicial);

        // Ambiente de execucao
        ContextoExecucaoImperativa2 ambiente =
                new ContextoExecucaoImperativa2(new ListaValor());

        // Programa
        Programa prog = new Programa(blocoCompleto);

        // Executa — esperado: excecao de variavel/identificador nao declarado
        executar(
            "{ var b = 3,\n" +
            "  proc escreveRecursivo (int a) {\n" +
            "    if (not (a == 0)) then {\n" +
            "      var x = 0; x := a - 1;\n" +
            "      write(\"Ola\");\n" +
            "      call escreveRecursivo(x)\n" +
            "    } else skip\n" +
            "  };\n" +
            "  call escreveRecursivo(a) }",
            prog,
            ambiente
        );
    }
}

File: Python/li2-python/Exemplo4.py
import traceback
from util.tipo_primitivo import TipoPrimitivo
from expression.exp_equals import ExpEquals
from expression.exp_not import ExpNot
from expression.exp_sub import ExpSub
from expression.id import Id
from expression.valor_inteiro import ValorInteiro
from expression.valor_string import ValorString
from exception.identificador_ja_declarado_exception import IdentificadorJaDeclaradoException
from exception.identificador_nao_declarado_exception import IdentificadorNaoDeclaradoException
from exception.variavel_ja_declarada_exception import VariavelJaDeclaradaException
from exception.variavel_nao_declarada_exception import VariavelNaoDeclaradaException
from command.atribuicao import Atribuicao
from command.comando_declaracao import ComandoDeclaracao
from command.if_then_else import IfThenElse
from command.sequencia_comando import SequenciaComando
from command.skip import Skip
from command.write import Write
from declaration.declaracao_composta import DeclaracaoComposta
from declaration.declaracao_variavel import DeclaracaoVariavel
from memory.contexto_compilacao_imperativa import ContextoCompilacaoImperativa
from memory.lista_valor import ListaValor
from command.chamada_procedimento import ChamadaProcedimento
from command.lista_expressao import ListaExpressao
from declaration.declaracao_parametro import DeclaracaoParametro
from declaration.declaracao_procedimento import DeclaracaoProcedimento
from declaration.def_procedimento import DefProcedimento
from declaration.lista_declaracao_parametro import ListaDeclaracaoParametro
from memory.contexto_execucao_imperativa2 import ContextoExecucaoImperativa2
from exception.procedimento_nao_declarado_exception import ProcedimentoNaoDeclaradoException
from programa import Programa

def executar(descricao, prog, ambiente):
    print("-" * 50)
    print("Programa: \n" + descricao + "\n")
    try:
        bem_tipado = prog.checaTipo(ContextoCompilacaoImperativa(ListaValor()))
        print("Bem tipado:", bem_tipado)
        if bem_tipado:
            saida = prog.executar(ambiente)
            print("Saida: ", end="")
            while saida is not None and saida.length() > 0:
                print(saida.getHead(), end=" ")
                saida = saida.getTail()
            print()
        else:
            print("Erro de tipo: execucao abortada.")
    except VariavelNaoDeclaradaException as e:
        print("[ERRO] Variavel nao declarada:", e)
    except VariavelJaDeclaradaException as e:
        print("[ERRO] Variavel ja declarada:", e)
    except ProcedimentoNaoDeclaradoException as e:
        print("[ERRO] Procedimento nao declarado:", e)
    except IdentificadorNaoDeclaradoException as e:
        print("[ERRO] Identificador nao declarado:", e)
    except IdentificadorJaDeclaradoException as e:
        print("[ERRO] Identificador ja declarado:", e)
    except Exception as e:
        print("[ERRO] Erro inesperado:", e)
    print("-" * 50)

if __name__ == "__main__":
    id_b = Id("b")
    id_a = Id("a")
    id_x = Id("x")
    id_escreve_recursivo = Id("escreveRecursivo")

    decl_b = DeclaracaoVariavel(id_b, ValorInteiro(3))
    param_a = DeclaracaoParametro(id_a, TipoPrimitivo.inteiro())
    lista_param_a = ListaDeclaracaoParametro(param_a)

    condicao = ExpNot(ExpEquals(id_a, ValorInteiro(0)))
    decl_x = DeclaracaoVariavel(id_x, ValorInteiro(0))
    atrib_x = Atribuicao(id_x, ExpSub(id_a, ValorInteiro(1)))
    write_ola = Write(ValorString("Ola"))
    chamada_recursiva = ChamadaProcedimento(id_escreve_recursivo, ListaExpressao(id_x))

    seq_then = SequenciaComando(atrib_x, SequenciaComando(write_ola, chamada_recursiva))
    bloco_then = ComandoDeclaracao(decl_x, seq_then)

    if_cmd = IfThenElse(condicao, bloco_then, Skip())
    def_escreve_recursivo = DefProcedimento(lista_param_a, if_cmd)
    decl_escreve_recursivo = DeclaracaoProcedimento(id_escreve_recursivo, def_escreve_recursivo)

    declaracoes = DeclaracaoComposta(decl_b, decl_escreve_recursivo)
    chamada_inicial = ChamadaProcedimento(id_escreve_recursivo, ListaExpressao(id_a)) # id_a is undeclared here, will raise error
    bloco_completo = ComandoDeclaracao(declaracoes, chamada_inicial)

    ambiente = ContextoExecucaoImperativa2(ListaValor())
    prog = Programa(bloco_completo)

    executar(
        "{ var b = 3,\n" +
        "  proc escreveRecursivo (int a) {\n" +
        "    if (not (a == 0)) then {\n" +
        "      var x = 0; x := a - 1;\n" +
        "      write(\"Ola\");\n" +
        "      call escreveRecursivo(x)\n" +
        "    } else skip\n" +
        "  };\n" +
        "  call escreveRecursivo(a) }",
        prog,
        ambiente
    )
